src/txn/txn_ckpt.c

```diff
@@ -862,7 +862,7 @@ __checkpoint_prepare(WT_SESSION_IMPL *session, bool *trackingp, const char *cfg[
             /* A checkpoint should never proceed when timestamps are out of order. */
             if (__wt_atomic_loadbool(&txn_global->has_oldest_timestamp) &&
```

Comment:
`WT_ASSERT_ALWAYS` runs in both release and diagnostic mode, so once we enter this branch `WT_ASSERT_ALWAYS(false)` means WiredTiger wil abort.
This makes the unlock redundant. I reckon we drop it and simplify the logic to

```
if (__wt_atomic_loadbool(&txn_global->has_oldest_timestamp)) {
    WT_ASSERT_ALWAYS(session,
    txn_global->oldest_timestamp <= txn_global->stable_timestamp,
    "oldest timestamp %s must not be later than stable timestamp %s when taking a "
    "checkpoint",
    __wt_timestamp_to_string(txn_global->oldest_timestamp, ts_string[0]),
    __wt_timestamp_to_string(txn_global->stable_timestamp, ts_string[1]));
}
```

Reply:
I am partially against this as if we ever change the behaviour of `WT_ASSERT_ALWAYS` (or remove it?), one may forget to release the lock.

Reply:
Discussed this offline. We'll stick with the current implementation.